File: backend/src/main/java/cn/etarch/mao/harness/mcp/preference/service/UserMcpPreferenceService.java
package cn.etarch.mao.harness.mcp.preference.service;

import cn.etarch.mao.harness.mcp.preference.entity.UserMcpPreference;
import cn.etarch.mao.harness.mcp.preference.mapper.UserMcpPreferenceMapper;
import com.baomidou.mybatisplus.core.conditions.query.LambdaQueryWrapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.List;

/**
 * 用户级 MCP 服务器启用偏好。
 * 语义：无记录 = 未单独配置，跟随管理后台全局启用状态。
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class UserMcpPreferenceService {

    private final UserMcpPreferenceMapper preferenceMapper;

    /**
     * 查询用户明确停用的 MCP 服务器 ID 列表（enabled=0）。
     */
    public List<Long> getDisabledServerIds(Long userId) {
        if (userId == null) {
            return List.of();
        }
        LambdaQueryWrapper<UserMcpPreference> qw = new LambdaQueryWrapper<>();
        qw.eq(UserMcpPreference::getUserId, userId)
                .eq(UserMcpPreference::getEnabled, 0);
        return preferenceMapper.selectList(qw).stream()
                .map(UserMcpPreference::getServerId)
                .toList();
    }

    /**
     * 查询用户对某台服务器的偏好；无记录返回 null（跟随全局）。
     */
    public UserMcpPreference get(Long userId, Long serverId) {
        return preferenceMapper.selectOne(
                new LambdaQueryWrapper<UserMcpPreference>()
                        .eq(UserMcpPreference::getUserId, userId)
                        .eq(UserMcpPreference::getServerId, serverId)
                        .last("LIMIT 1"));
    }

    /**
     * 保存用户级启用/停用偏好。
     * enabled=true 时删除记录（无记录 = 跟随全局启用，语义最简）；
     * enabled=false 时 upsert enabled=0。
     */
    public void save(Long userId, Long serverId, boolean enabled) {
        if (userId == null || serverId == null) {
            return;
        }
        UserMcpPreference existing = get(userId, serverId);
        if (enabled) {
            if (existing != null) {
                // 删除记录 = 未单独配置，跟随全局启用状态
                preferenceMapper.deleteById(existing);
                log.info("Cleared MCP preference: userId={}, serverId={} (follows global)", userId, serverId);
            }
            return;
        }
        if (existing == null) {
            UserMcpPreference row = new UserMcpPreference();
            row.setUserId(userId);
            row.setServerId(serverId);
            row.setEnabled(0);
            preferenceMapper.insert(row);
        } else {
            existing.setEnabled(0);
            preferenceMapper.updateById(existing);
        }
        log.info("Saved MCP preference: userId={}, serverId={}, enabled=false", userId, serverId);
    }

    /** 查询用户的全部偏好记录（含启用与停用），供设置页初始化显示。 */
    public List<UserMcpPreference> listByUser(Long userId) {
        if (userId == null) {
            return List.of();
        }
        return preferenceMapper.selectList(
                new LambdaQueryWrapper<UserMcpPreference>()
                        .eq(UserMcpPreference::getUserId, userId));
    }

    /**
     * 删除指向指定服务器的全部偏好记录（服务器被删除时级联清理）。
     */
    public void deleteByServer(Long serverId) {
        if (serverId == null) {
            return;
        }
        preferenceMapper.delete(
                new LambdaQueryWrapper<UserMcpPreference>()
                        .eq(UserMcpPreference::getServerId, serverId));
        log.info("Cleared MCP preferences for serverId={}", serverId);
    }
}

File: backend-ts/src/harness/mcp/preference/service/user-mcp-preference.service.ts
import { harnessLog } from '../../../log.js';
import type { UserMcpPreferenceMapper } from '../../mapper/mcp-server.mapper.js';
import type { UserMcpPreference } from '../../entity/mcp-server.js';

export class UserMcpPreferenceService {
  constructor(private readonly preferenceMapper: UserMcpPreferenceMapper) {}

  async getDisabledServerIds(userId: number | null | undefined): Promise<number[]> {
    if (userId == null) return [];
    const rows = await this.preferenceMapper.listDisabledServerIds(userId);
    return rows.map((r) => r.serverId);
  }

  get(userId: number, serverId: number): Promise<UserMcpPreference | null> {
    return this.preferenceMapper.get(userId, serverId);
  }

  async save(userId: number | null, serverId: number | null, enabled: boolean): Promise<void> {
    if (userId == null || serverId == null) return;
    const existing = await this.get(userId, serverId);
    if (enabled) {
      if (existing?.id != null) {
        await this.preferenceMapper.deleteById(existing.id);
        harnessLog('info', `Cleared MCP preference: userId=${userId}, serverId=${serverId} (follows global)`);
      }
      return;
    }
    if (existing == null) {
      await this.preferenceMapper.insert({ userId, serverId, enabled: 0 });
    } else if (existing.id != null) {
      await this.preferenceMapper.updateById(existing.id, { enabled: 0 });
    }
    harnessLog('info', `Saved MCP preference: userId=${userId}, serverId=${serverId}, enabled=false`);
  }

  listByUser(userId: number | null): Promise<UserMcpPreference[]> {
    if (userId == null) return Promise.resolve([]);
    return this.preferenceMapper.listByUser(userId);
  }

  async deleteByServer(serverId: number | null): Promise<void> {
    if (serverId == null) return;
    await this.preferenceMapper.deleteByServer(serverId);
    harnessLog('info', `Cleared MCP preferences for serverId=${serverId}`);
  }
}
